### ISSUE TITLE
Feature: Add core hooks for altering how death/resurrection works.

### ISSUE THREAD

**Issue by Viskus** [azerothcore/azerothcore-wotlk#17370]
### Describe your feature request or suggestion in detail

Add hooks to change or alter how death works.
As hardcore servers are becoming more popular it would be great to have the option to keep someone as a ghost perm.
The hook could potentially change how the resurrect button works and make that customisable where it may not change the text but the fuction would be different, example would be that pressing the button would just teleport the character back to the graveyard.

### Describe a possible solution to your feature or suggestion in detail

If the hooks are present in the core then we can change how it works with modules hopefully.
I'm not technically minded in this area but hoping that someone could see some value in this feature!

**Comment by kissingers** [comment]
Yes, also at LUAif can add this support, so when those died can appear a diag to select or change the onelife died way.
now only can add a NPC when die.

**Comment by netweaver** [comment]
i am working on it . both in ac and eluna.
AC side:
1. add sScriptMgr->OnBeforePlayerResurrect at the begining of Player::ResurrectPlayer
2. define OnBeforePlayerResurrect in PlayerScript.cpp   , call playerscritp :: OnBeforePlayerResurrect

eluna side:
1.  define Eluna_PlayerScript : OnBeforePlayerResurrect , call Eluna:: OnBeforeResurrect
2.  define Eluna:: OnBeforeResurrect , call PLAYER_EVENT_BEFORE_RESURRECT Event handlers
3.  add PLAYER_EVENT_BEFORE_RESURRECT  = 59  in enum PlayerEvents

it's under testing, PR soon ,maybe after national day holidays
